src/System.Device.Gpio/System/Device/I2c/I2cBus.cs

```diff
@@ -0,0 +1,83 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace System.Device.I2c
+{
+    /// <summary>
+    /// I2C bus communication channel.
+    /// </summary>
+    public abstract partial class I2cBus : IDisposable
+    {
+        /// <summary>
+        /// Creates default I2cBus
+        /// </summary>
+        /// <param name="busId">The bus ID.</param>
+        /// <returns>I2cBus instance.</returns>
+        public static I2cBus Create(int busId)
+        {
+            if (Environment.OSVersion.Platform == PlatformID.Win32NT)
+            {
+                return CreateWindows10I2cBus(busId);
+            }
+            else
+            {
+                return UnixI2cBus.Create(busId);
+            }
+        }
+
+        /// <summary>
+        /// Creates I2C device.
+        /// </summary>
+        /// <param name="deviceAddress">Device address related with the device to create.</param>
+        /// <returns>I2cDevice instance.</returns>
+        public abstract I2cDevice CreateDevice(int deviceAddress);
+
+        /// <summary>
+        /// Removes I2C device.
+        /// </summary>
+        /// <param name="deviceAddress">Device address to create</param>
+        public abstract void RemoveDevice(int deviceAddress);
+
+        /// <summary>
+        /// Reads data from the specified device.
+        /// </summary>
+        /// <param name="deviceAddress">Device address to read from.</param>
+        /// <param name="buffer">Buffer to read data to.</param>
+        public abstract void Read(int deviceAddress, Span<byte> buffer);
```

Comment:
Do you mind explaining why you need Read, Write and WriteRead methods on the Bus? I had envisioned the Bus being more of a I2cDevice manager, that you would use to create and remove devices, and then use the devices you got back to actually perform the read, write and writeRead operations. WIth this now you have read, write and writeRead both on the bus and on the device.

Reply:
Basically my comment here is, the way this PR is today, about 80% of the API in i2cBus is also present on I2cDevice, so it would be good to separate a bit more the responsabilities of  both. Today, I2cDevice class can Create i2c devices, Read, Write and Dispose them. I2cBus with your changes can do exactly that as well, plus  manage i2c Devices (Remove devices). Why not make I2cBus all about I2cDevice management (possibly even consider deprecating I2cDevice.Create so folks try to use I2cBus.CreateDevice instead) and leave the read/write operations on the device?

Reply:
i.e.: I2C scanner, sending reset to I2C Bus, devices using multiple I2C addresses (see the issue in the description)

Reply:
GpioController is different as we don't have a GpioPin yet, so this is not an issue there. My concern is that  we will now have 2 classes that are more than 80% the same. It would be better to separate responsabilities. From the issue seems like having read/write/readwrite on bus is just  a convinience thing, I think we should prioritize having separate concepts for each class over convinience.

Reply:
I thought the same, but I can see the use cases @krwq points out. Could we at least have the read/write/writeread methods as extension methods somewhere else (doesn't even need to be in System.Devices)?

Reply:
The extension methods won't be a good idea. I.e. in case of Ft4222 or Unix implementation they only have a concept of a bus and I2cDevice is not present at all, in case of Windows implementation they don't have a concept of a bus. In case of Windows 10 extension method would need to create a device just for the sake of doing this operation while in the case of Unix/Ft4222 you create a device for the sake of having device. I do not feel strongly about having these on the bus but it feels to me like depending on the device either one or the other one will make more sense

Reply:
I have the impression we got a clash of partially contradicting concepts (or requirements) here. What you are implementing here is a `I2cBus_Driver_`, while what I was more expecting is a `I2cBus_Manager_`, that would be similar in use to the GpioController. 

Your requirements/implementation:
- Implement a hardware-dependent component that is used to talk to devices, actually implementing the read/write operations.
- Create either a bus or a device to forward to a binding, whichever is more convenient

My requirements/expectations:
- The I2cBus manages devices, but is not hardware-dependent, that part is left to the devices
- Typically, I2cBus is a singleton instance. Only one instance per physical bus exists
- Supports per-bus operations, such as scanning or resetting, using the driver for this purpose
- I2cBus shall be used to create devices

I guess we'll have to find consensus on which path to go, maybe what we really need is an `I2cBus` (as management instance) and an `I2cBusDriver` (as mostly what this PR proposes). I would not mix the two.

Reply:
Another element in favor of the read and write on the I2cBus is support for SMBus which is very close to I2C with alerting and few other elements.
I agree that it can be a bit confusing having both read and write on both the I2cDevice and the I2cBus but it's a matter of scenario. And so far, we only allow devices in bindings having I2cDevice and we should continue like this except for specific scenarios like sense hat where having a bus will make sense but is not necessary needed.
Documentation and guidelines and PR review should be strict about this.

Reply:
I'm not opposing the creation of I2cBus, but I think the proposed approach needs a conceptual review. I believe this is mostly a move of functions from one class to another, but does not provide the abstraction we really want. As I noted above, I think the requirements are not clear.

Reply:
@pgrawehr what would be then the classes abstract for what you describe (even if not complete) to have a better understanding of the relations between all of them?

Reply:
@Ellerbach : I hope I'll be able to provide a schema of what I mean by tomorrow.

Reply:
![I2cBusPR](https://user-images.githubusercontent.com/13809874/102238744-673f2300-3ef6-11eb-879b-f99182bf6112.png)
![I2cBusWithDriver](https://user-images.githubusercontent.com/13809874/102238763-6c9c6d80-3ef6-11eb-91eb-736b465f68e9.png)

The first schema shows how things look with this PR, the second schema is somewhat what I would prefer. For further extension there should be a separation between adding features that affect devices or busses in general (such as bus scanning, resetting, ref-counting devices, writing exactly 7 bytes or whatever) and the actual I2C low-level transport implementation. 

Rationale: Extension will be different for the two cases, and often you wouldn't want to do both at the same time. I.e. for the board PR, I'll be adding pin and (eventually) device managing capabilities, which I could add by inheriting from I2cBus and/or I2cDevice. For adding a bus scanning function, one could inherit from I2cBus, or even write it as an extension function. However, for the Arduino PR, I'll be writing an I2c Driver, but leave the bus and device classes probably untouched.

Reply:
So if I summarize correctly, you propose to:

- Add another layer I2cBusDriver
- Create either the I2cBus either I2cDevice from the I2cBusDriver

So if I understand correctly that will imply:

- To have the I2cBusDriver exposed => rather than have I2cBus and I2cDevice exposed only, we will have then the 3. I guess that would bring even more confusions

So I would prefer to go for I2cBus and I2cDevice exposed only and hide some of the plumbing behind. Some systems have a notion of bus (Linux and in general all systems like FT4222, Arduino, etc), only exception is Windows (so far). So the I2cBus seems more suitable to create instances of I2cDevice than having an intermediate driver.

I may have miss something in your proposal @pgrawehr

Reply:
@Ellerbach Yes, that sums up my proposal quite well. I think however, that the additional publicly visible interface would cause very little harm, because most users won't have to deal with it, similar to the existing GpioDriver abstract base. However, it would make deriving instances for the board, FT4222, Arduino etc, much more clean. We should talk this over in tomorrow's call.